streams/src/test/java/org/apache/kafka/streams/kstream/internals/CogroupedKStreamImplTest.java

```diff
@@ -0,0 +1,369 @@
+    private static final String TOPIC = "topic";
+    private final StreamsBuilder builder = new StreamsBuilder();
+    private KGroupedStream<String, String> groupedStream;
+    private CogroupedKStream<String, String> cogroupedStream;
+
+    private final Properties props = StreamsTestUtils
+        .getStreamsConfig(Serdes.String(), Serdes.String());
+
+    private static final Aggregator<String, String, String> STRING_AGGREGATOR = (key, value, aggregate) ->
+        aggregate + value;
+
+    private static final Initializer<String> STRING_INITIALIZER = () -> "";
+
+    private static final Aggregator<String, String, Integer> STRSUM_AGGREGATOR = (key, value, aggregate) ->
+        aggregate + Integer.parseInt(value);
+
+    private static final Aggregator<? super String, ? super Integer, Integer> SUM_AGGREGATOR = (key, value, aggregate) ->
+        aggregate + value;
+
+    private static final Initializer<Integer> SUM_INITIALIZER = () -> 0;
+
+
+    @Before
+    public void setup() {
+        final KStream<String, String> stream = builder.stream(TOPIC, Consumed
+            .with(Serdes.String(), Serdes.String()));
+
+        groupedStream = stream.groupByKey(Grouped.with(Serdes.String(), Serdes.String()));
+        cogroupedStream = groupedStream.cogroup(MockAggregator.TOSTRING_ADDER);
+    }
+
+    @Test(expected = NullPointerException.class)
+    public void shouldThrowNPEInCogroupIfKGroupedStreamIsNull() {
+        cogroupedStream.cogroup(null, MockAggregator.TOSTRING_ADDER);
+    }
+
+    @Test(expected = NullPointerException.class)
+    public void shouldNotHaveNullAggregatorOnCogroup() {
+        cogroupedStream.cogroup(groupedStream, null);
+    }
+
+    @Test(expected = NullPointerException.class)
+    public void shouldNotHaveNullInitializerOnAggregate() {
+        cogroupedStream.aggregate(null, Materialized.as("store"));
+    }
+
+    @Test(expected = NullPointerException.class)
+    public void shouldNotHaveNullMaterMaterializedOnAggregate() {
+        cogroupedStream.aggregate(STRING_INITIALIZER, null);
+    }
+    
+
+    @Test
+    public void shouldCoGroupAndAggregateSingleKStreams() {
+
+        final KStream<String, String> test = builder.stream("one", stringConsumed);
+
+        final KGroupedStream<String, String> groupedOne = test.groupByKey();
+
+        final KTable<String, String> customers = groupedOne.cogroup(STRING_AGGREGATOR)
+                .aggregate(STRING_INITIALIZER);
+
+        customers.toStream().process(processorSupplier);
+
+        try (final TopologyTestDriver driver = new TopologyTestDriver(builder.build(), props)) {
+            final TestInputTopic<String, String> testInputTopic = driver.createInputTopic("one", new StringSerializer(), new StringSerializer());
+            testInputTopic.pipeInput("k1", "A", 0);
+            testInputTopic.pipeInput("k11", "A", 0);
+            testInputTopic.pipeInput("k11", "A", 0);
+            testInputTopic.pipeInput("k1", "A", 0);
```

Comment:
Should we add different timestamps? Also include out-of-order data to check that output timestamp is computed correctly?

Reply:
There are other tests that use different timestamps and out of order. This is making sure it handles all cases